### PR TITLE
Issue #68: Redirection logic for auth

### LINKED ISSUE #68: Triple check logic around tool_mfa_after_require_login
The entire security of this plugin hinges on this. This needs to be bulletproof.

I've scanned require_login for all the weird edge cases that happen before this hook fires:

* [x] Add unit tests with a data provider with urls. Need to make sure you can't craft a url to get into some other page by hacking query params, or file arguments etc

* [x] I would also like special checks for non web things, ajax calls etc. We need to allow enough stuff to work to allow the page to render fully, but you shouldn't be able to call say an arbitrary ajax web service yet.

* [x] test for when a user logs in the first time, but is not fully setup (eg gets redirected to /user/edit.php) BEFORE this hook is called

* [x] test guest access

* [x] test for forced password change logic. this fires BEFORE this hook

* [x] site policy core redirect fires BEFORE

* [x] soft maintenance mode check

* [x] login as checks, see #22

* [x] enrolment plugin redirects

* [x] if we are in forced lockdown, AND you end up on one of these pages where core has redirected, then we cannot redirect to the auth page or we will go in circles. We need to just print an error and die.

### PR DESCRIPTION
**Pull request by Peterburnett**
Closes #68,  #108, #107 and #36.

### CODE REVIEW

File: classes/manager.php [hunk 1]
```diff
@@ -257,12 +261,101 @@ public static function set_pass_state() {
...
+            return 0;
+        }
+
+        // WS/AJAX check.
+        if (WS_SERVER || AJAX_SCRIPT) {
+            return self::REDIRECT_EXCEPTION;
+        }
+
+        // User/edit.php.
+        $useredit = new \moodle_url('/user/edit.php');
```
Comment (brendanheywood) [inline classes/manager.php]: I think this needs to be more like if (user_not_full_setup() then NO_REDIRECT

Test this in profile editing when some ajax calls

File: classes/manager.php [hunk 2]
```diff
@@ -257,12 +261,101 @@ public static function set_pass_state() {
...
+        }
+
+        // Site policy.
+        if (isset($USER->policyagreed) && !$USER->policyagreed &&
+            defined('NO_SITEPOLICY_CHECK') && !NO_SITEPOLICY_CHECK) {
+            return self::NO_REDIRECT;
+        }
+
+        // Soft maintenance mode.
+        if (!empty($CFG->maintenance_enabled)) {
```
Comment (brendanheywood) [inline classes/manager.php]: I'd move this one to the top

I think there is also a couple other similar edge cases to test / check during install / upgrade:

if (isset($CFG->adminsetuppending)) {

if (during_initial_install())

if ($CFG->upgraderunning

Some of these might be redundant and overlap

File: lib.php [hunk 3]
```diff
@@ -44,10 +45,12 @@ function tool_mfa_after_require_login() {
             $SESSION->tool_mfa_setwantsurl = true;
```
Comment (brendanheywood) [inline lib.php]: I think only do this if you actually redirect, so move into the if below?

File: lib.php [hunk 4]
```diff
@@ -31,8 +31,9 @@
  * @return void
  * @throws \moodle_exception
  */
-function tool_mfa_after_require_login() {
-    global $SESSION, $ME, $CFG;
+function tool_mfa_after_require_login($courseorid = null, $autologinguest = null, $cm = null,
+    $setwantsurltome = null, $preventredirect = null) {
```
Comment (brendanheywood) [inline lib.php]: I've never actually used $setwantsurltome but seems like we should honor this flag too?

File: classes/manager.php [hunk 5]
```diff
@@ -257,12 +261,115 @@ public static function set_pass_state() {
...
+
+        // Initial installation.
+        // We get this for free from get_plugins_with_function.
+
+        // Upgrade check.
+        // We get this for free from get_plugins_with_function.
+
+        // Honor prevent_redirect.
+        if ($preventredirect) {
+            return 0;
```
Comment (brendanheywood) [inline classes/manager.php]: is this intentional?
